advanced_source/static_quantization_tutorial.py

```diff
@@ -0,0 +1,621 @@
+        if cnt >= ntrain_batches:
+            print('Loss', avgloss.avg)
+
+            print('Training: * Acc@1 {top1.avg:.3f} Acc@5 {top5.avg:.3f}'
+                  .format(top1=top1, top5=top5))
+            return
+
+    print('Full imagenet train set:  * Acc@1 {top1.global_avg:.3f} Acc@5 {top5.global_avg:.3f}'
+          .format(top1=top1, top5=top5))
+    return
+
+######################################################################
+# We fuse modules as before
+
+qat_model = load_model(saved_model_dir + float_model_file)
+qat_model.fuse_model()
+
+optimizer = torch.optim.SGD(qat_model.parameters(), lr = 0.0001)
+qat_model.qconfig = torch.quantization.get_default_qat_qconfig('fbgemm')
+
+######################################################################
+# Finally, ``prepare_qat`` performs the "fake quantization", preparing the model for quantization-aware
+# training
+
+torch.quantization.prepare_qat(qat_model, inplace=True)
+print('Inverted Residual Block: After preparation for QAT, note fake-quantization modules \n',qat_model.features[1].conv)
+
+######################################################################
+# Training a quantized model with high accuracy requires accurate modeling of numerics at
+# inference. For quantization aware training, therefore, we modify the training loop by:
+#
+# - Switch batch norm to use running mean and variance towards the end of training to better
+#   match inference numerics.
+# - We also freeze the quantizer parameters (scale and zero-point) and fine tune the weights.
+
+num_train_batches = 20
+
+# Train and check accuracy after each epoch
+for nepoch in range(8):
+    train_one_epoch(qat_model, criterion, optimizer, data_loader, torch.device('cpu'), num_train_batches)
+    if nepoch > 3:
+        # Freeze quantizer parameters
+        qat_model.apply(torch.quantization.disable_observer)
+    if nepoch > 2:
+        # Freeze batch norm mean and variance estimates
+        qat_model.apply(torch.nn.intrinsic.qat.freeze_bn_stats)
+
+    # Check the accuracy after each epoch
+    quantized_model = torch.quantization.convert(qat_model.eval(), inplace=False)
+    quantized_model.eval()
+    top1, top5 = evaluate(quantized_model,criterion, data_loader_test, neval_batches=num_eval_batches)
+    print('Epoch %d :Evaluation accuracy on %d images, %2.2f'%(nepoch, num_eval_batches * eval_batch_size, top1.avg))
+
+#####################################################################
+# More on quantization-aware training:
+#
+# - QAT is a super-set of post training quant techniques that allows for more debugging.
+#   For example, we can analyze if the accuracy of the model is limited by weight or activation
+#   quantization.
+# - We can also simulate the accuracy of a quantized model in floating point since
+#   we are using fake-quantization to model the numerics of actual quantized arithmetic.
+# - We can mimic post training quantization easily too.
```

Comment:
We can also state that quantization aware training yields an accuracy of over 71% on the entire imagenet dataset, which is close to the floating point accuracy of 71.9%. In addition, we can also state that the simple quantization technique that we first try gives us 63% accuracy and the per-channel technique boosts it to 67%